flink-runtime/src/main/java/org/apache/flink/runtime/state/proxy/ProxyKeyedStateBackend.java

```diff
@@ -0,0 +1,306 @@
+import org.apache.flink.runtime.state.AbstractKeyedStateBackend;
+import org.apache.flink.runtime.state.CheckpointStreamFactory;
+import org.apache.flink.runtime.state.KeyGroupedInternalPriorityQueue;
+import org.apache.flink.runtime.state.Keyed;
+import org.apache.flink.runtime.state.KeyedStateFunction;
+import org.apache.flink.runtime.state.KeyedStateHandle;
+import org.apache.flink.runtime.state.PriorityComparable;
+import org.apache.flink.runtime.state.SnapshotResult;
+import org.apache.flink.runtime.state.StateSnapshotTransformer;
+import org.apache.flink.runtime.state.heap.HeapKeyedStateBackend;
+import org.apache.flink.runtime.state.heap.HeapPriorityQueueElement;
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.IOUtils;
+
+import javax.annotation.Nonnull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.RunnableFuture;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** */
+public class ProxyKeyedStateBackend<K> extends AbstractKeyedStateBackend<K> {
+    // wrapped keyed state backend, either HeapKeyedStateBackend or RocksDBKeyedStateBackend
+    AbstractKeyedStateBackend<K> keyedStateBackend;
+
+    private static final Map<Class<? extends StateDescriptor>, StateFactory> STATE_FACTORIES =
+            Stream.of(
+                            Tuple2.of(
+                                    ValueStateDescriptor.class,
+                                    (StateFactory) ProxyValueState::create),
+                            Tuple2.of(
+                                    ListStateDescriptor.class,
+                                    (StateFactory) ProxyListState::create),
+                            Tuple2.of(
+                                    ReducingStateDescriptor.class,
+                                    (StateFactory) ProxyReducingState::create),
+                            Tuple2.of(
+                                    AggregatingStateDescriptor.class,
+                                    (StateFactory) ProxyAggregatingState::create),
+                            Tuple2.of(
+                                    MapStateDescriptor.class, (StateFactory) ProxyMapState::create))
+                    .collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+
+    public ProxyKeyedStateBackend(AbstractKeyedStateBackend<K> keyedStateBackend) {
+        super(
+                keyedStateBackend.kvStateRegistry,
+                keyedStateBackend.keySerializer,
+                keyedStateBackend.userCodeClassLoader,
+                keyedStateBackend.executionConfig,
+                keyedStateBackend.ttlTimeProvider,
+                keyedStateBackend.cancelStreamRegistry,
+                keyedStateBackend.keyGroupCompressionDecorator,
+                keyedStateBackend.keyContext);
+        this.keyedStateBackend = keyedStateBackend;
```

Comment:
In runtime, we end up with 4 backend:
```
Proxy extends Abstract (has it's own fields AND refers to passed fields)
    Proxied extends Abstract (has it's own fields AND refers to passed fields)
```
Right?

I see the following drawbacks:
1. Double caching and lookup in `keyValueStatesByName`
2. Error-prone: 4 backend classes refer to the same `kvStateRegistry`, `keySerializer` etc. making it easy to either call some method twice or not call it for a wrapped backend (e.g. `setCurrentKey` isn't called for the wrapped backend)
4. To me this hierarchy is very confusing

Why not refactor this in some way or another? Can `AbstractKeyedStateBackend` be also a proxy (i.e. replace inheritance with delegation)?

Reply:
All these questions are great! For all the reasons mentioned or not mentioned above, I feel this hierarchy is confusing as well. So I've spent some time during the weekend simplifying it. 

Originally, the reason I extend `AbstractKeyedStateBackend` is I thought all fields have to be wrapped and delegated. So if I implement `CheckpointableKeyedStateBackend`, I basically rewrite everything in `AbstractKeyedStateBackend`. 

But it is not true, only the InternalKvState related (that's the state representation to internal ops) have to be wrapped, everything else (keys, namespace e.t.c) is kept untouched.

So, I reimplement this to  `implements CheckpointableKeyedStateBackend`. With a bit more duplicated code, I think this is much more simple and clean.

**However**, the problem `keyValueStatesByName` is duplicated but one instance that isn't actually used is still there. I am still not happy with this part, but much better.